### ISSUE TITLE
V1.1.0 computed breaks when computing a different shape for a computed signal

### ISSUE THREAD

**Issue by stanleyzapata** [cesarParra/lwc-signals#21]
This was working under v1.0.1 but breaks under v1.1.0

<img width="694" alt="Screenshot 2024-11-29 at 3 42 15 PM" src="https://github.com/user-attachments/assets/180c4e2f-1dd1-499f-9e4c-06be5b3663f3">

**Comment by cesarParra** [comment]
Thanks @stanleyzapata .

Can you provide a little bit more information if possible for me to replicate? I am looking for the following info:
* It seems like `availablePaymentTypes` is an array. Can you provide a sample of how this `signal` gets declared and how is it updated? For instance, is it being declared with the `track` option turned on, and when it gets updated, is it getting reassigned or does it get updated through an array function like a push?
* How is `availablePaymentTypeOptions` being used. Is this part of a class property or is it a standalone exported variable?

**Comment by stanleyzapata** [comment]
1. **`availablePaymentTypes`** is a computed signal derived from:
     - `configuredPaymentTypes`: A resource signal that fetches data from an Apex method and updates based on its dependencies.
     - `registeredPaymentTemplateByComponent`: A computed signal that reshapes `registeredPaymentComponents`, a tracked signal (an array of objects).

2. **Update Behavior**:
   - The tracked signal `registeredPaymentComponents` is updated via reassignment in a function, for example:
     ```javascript
     registeredPaymentComponents.value = [...registeredPaymentComponents.value, newItem];
     ```
   -Updates to `registeredPaymentComponents` propagate to its dependent computed signals, including `registeredPaymentTemplateByComponent` and `availablePaymentTypes`.

3. **`availablePaymentTypeOptions` Usage**:
   - `availablePaymentTypeOptions` is a standalone computed signal that transforms `availablePaymentTypes` into an array of objects with `{ label, value }`. 
   - It’s exported and used to provide options for dropdowns or other UI components.

**Comment by cesarParra** [comment]
Thanks for the added info @stanleyzapata .

Unfortunately it seems I am going to need some additional help reproducing this on my side. I stood up some LWCs to try and replicate based on your comment, but all transformations are working as expected on my side.

Here is the code that I created to try and reproduce. If you have the chance, can you take a look and see if there's something I'm missing?

https://github.com/cesarParra/lwc-signals/compare/track-computed-values?expand=1

Here is what I built:
![Kapture 2024-12-02 at 13 30 20](https://github.com/user-attachments/assets/71d9e8b4-7c57-43ee-8115-3f03869444cd)

There are 2 components + 1 service that holds all of the signals. The component on top displays what the signals/resources/computed values have at any given time. The component at the bottom let me register new "Payment Types" by their name, which then recompute the available payment type signal, which in turn recomputes the `options` signal.

**Comment by stanleyzapata** [comment]
I tried to replicate my issue with your code, and couldn't replicate it either. After debugging some more, I found that the issue in my code is being caused by a circular reference from an effect.

### What I Found

1. The issue is caused by an effect that sets a selected payment type based on `availablePaymentTypes`. The effect logic looks something like this:
   - It checks if there are `availablePaymentTypes`, and if there are, it performs a number of checks:
     - Verifies if the value in the `selectedPaymentType` signal is valid.
     - If `selectedPaymentType` doesn’t have a value, it sets the first value from `availablePaymentTypes`.
   - This creates a reference to `availablePaymentTypes` in the effect.

2. When `availablePaymentTypes` changes, it notifies its subscribers, including the effect. However:
   - During the notification phase, the effect attempts to access `availablePaymentTypes` again.
   - At this point, the code detects that `availablePaymentTypes` is still in a "computing" state, even though it has already finished computing and is simply notifying subscribers.

### Call Stack Tree

```
availablePaymentTypes (computed) 
  └── Notifies Subscribers
       └── selectedPaymentEffect
            └── Reads availablePaymentTypes (dependency)
                 └── Detects computing state, re-triggers availablePaymentTypes computation causes an exception that stops notifications to other subscribers
```

I’d appreciate any insights or suggestions on whether this behavior is expected and if there’s a better way to handle such cases.

**Comment by cesarParra** [comment]
It depends. If the effect reads and writes the same signal, then yes, that is a known limitation and that is expected to cause a circular reference/infinite loop*. Otherwise, if the effect just writes but does not read then that might be a bug.

For the first use case, I am planning on investigating adding something similar to the Preact `untracked` functionality (https://github.com/preactjs/signals?tab=readme-ov-file#untrackedfn) to be able to read a signal inside of an effect without subscribing to future notifications, but that is still in the idea stage.

*note: even if the effect doesn't read and write the signal in the same function, it is still possible a circular reference exists if the effects reads something else that ends up reading the signal. This will also cause issues, so the recommendation is to avoid writing inside of effects that might have dependencies, at least until that untracked functionality is implemented

If you are dealing with the second case, would it be possible for you to share a little bit of that code (or a recreation of that code)?

**Comment by stanleyzapata** [comment]
If I understood correctly my issue falls under the first case, the effect reads from availablePaymentTypes to check if there are available payment types and determine if the selectedPaymentType signal should be updated.

I don't think this should cause circular reference unless I update the availablePaymentTypes signal and seem more like a bug to me.

Here’s a simplified version of the code:
```javascript
$effect(function selectedPaymentEffect() {
    const availableTypes = availablePaymentTypes.value;

    if (availableTypes.length === 0) {
        if (selectedPaymentType.value !== null) {
            selectedPaymentType.value = null;
        }
        return;
    }

    let selectedPaymentTypeValue = selectedPaymentType.value;

    // If the selected payment type is not available, reset it
    if (selectedPaymentTypeValue && !availableTypes.find(pt => pt.id === selectedPaymentType.value.id)) {
        selectedPaymentTypeValue = null;
    }

    // If no payment type is selected, select the first one
    if (!selectedPaymentTypeValue) {
        selectedPaymentTypeValue = availableTypes[0];
    }

    // Set the selected payment type, if it has changed
    if (selectedPaymentType.value !== selectedPaymentTypeValue) {
        selectedPaymentType.value = selectedPaymentTypeValue;
    }
});
```

I also want to highlight that is a bit a problem that if a single subscriber has issues with an update it affects the other subscribers and stops the complete update chain

**Comment by cesarParra** [comment]
Given that code sample it looks like there are several places that might be problematic. Both `selectedPaymentType` and `selectedPaymentType` are being read and written in that function.

Specifically the logic for `selectedPaymentType` seems like it will cause issues no matter what. First it is reading the value, if it is null it it setting it, but it is also setting it when it is not null but different than the previous value, so it seems like it will always get set no matter what. 

I would recommend to favor computing the values you want, rather than manipulating signals in effects (similar to favor declarative programming over imperative). 

In this case, since you want to make sure that invalid payment types are not being selected, then I would recommend to avoid the consumer from manipulating the signal manually, and hide it behind an API that you control.

Here's a sample implementation that achives the same thing

```
const availablePaymentTypes = $signal([]);
const manuallySelectedPaymentType = $signal(undefined); // keep private
const selectedPaymentType = $computed(() => manuallySelectedPaymentType.value ?? availablePaymentTypes.value?.[0]); // ok to export

// ok to export
function selectPaymentType(paymentType) {
  if (availablePaymentTypes.value.includes(paymentType)) {
    manuallySelectedPaymentType.value = paymentType;
  }
  throw new Error('Invalid payment type'); // or handle it as you see fit
}

```

---

For the exception handling, letting the error throw is a decision made on purpose (other Signal libraries seem to do the same thing as well). If we catch all errors then there might be a perception that things are fine and were calculated correctly, when in fact some issues occurred. Another option is to have an `error` property in the signal but then we would force the user to have to check every effect and computed for errors, which would make the API very complicated to use. And we can't let some values succeed and others fail because JS is single-threaded, so everything occurs in the same transaction.

The decision for which errors to handle is left to the developer. If you have computed/effects that should not stop execution then you can wrap the function in a try/catch and decide how your application should respond for your particular use case

**Comment by stanleyzapata** [comment]
Here’s a refined version of your response:

---

Oh, I didn’t think about doing it that way! If I follow the approach you mentioned, I believe it would work well to solve my case.

What I was trying to explain earlier is that errors in my previous implementation weren’t surfacing properly. If I used the logic I mentioned earlier, the errors seemed to get swallowed. I had to add debug logs in the library and step through the code multiple times to figure out what was happening. It seems that when one of the subscribers fails (in this case, the `$effect` subscriber), the error gets caught by the computed signal, and at some point, it gets lost. As a result, the `computedGetter` never runs. I suspect this might have something to do with `$computed` relying on `$effect` under the hood.

**Comment by stanleyzapata** [comment]
Oh, I didn’t think about doing it that way! If I follow the approach you mentioned, I believe it would work well to solve my case.

What I was trying to explain earlier is that errors in my previous implementation weren’t surfacing properly. If I used the logic I mentioned earlier, the errors seemed to get swallowed. I had to add debug logs in the library and step through the code multiple times to figure out what was happening. It seems that when one of the subscribers fails (in this case, the `$effect` subscriber), the error gets caught by the computed signal, and at some point, it gets lost. As a result, the `computedGetter` never runs. I suspect this might have something to do with `$computed` relying on `$effect` under the hood.
